### ISSUE TITLE
Add Authentication parameters

### ISSUE THREAD

**Issue by beescuit** [SwitchbladeBot/bladey-api#6]
An authentication header is necessary for the DB-related endpoints so that we can modify the user's balance for example

**Comment by dsgpaiva** [comment]
but *how*

**Comment by pedrofracassi** [comment]
Using an Authentication header...?

**Comment by dsgpaiva** [comment]
but what's the header going to match, like
+ the user's OAuth2 token?
+ a special token that you'll make? (although that would be stupid sending that over a open-source website)

**Comment by pedrofracassi** [comment]
We'd send the user's OAuth token. There's no problem with open source: Our machines are private and no tokens will be exposed on the repo.

**Comment by dsgpaiva** [comment]
And how'd you store that in the database w/o showing its login details?

**Comment by pedrofracassi** [comment]
We don't have to. We'd make a request to discord to check if the header is valid. No need to store it.

**Comment by dsgpaiva** [comment]
wait for real? that's a thing?

**Comment by dsgpaiva** [comment]
but like, it can be forged, unless CORS is used

**Comment by dsgpaiva** [comment]
aaaaaah, so that's why it's installed
nvm

**Comment by pedrofracassi** [comment]
"Can be forged", just like I can forge a request to google using my own token

**Comment by beescuit** [comment]
Umm... Why don't we use a fixed key list..? There's no need to authenticate with an user account as this is more likely to be used by the server side in some specific "administration" endpoints

**Comment by dsgpaiva** [comment]
and why should we? besides, how can we, like, send one of the keys through a static, open-source website?

**Comment by pedrofracassi** [comment]
@beescuit If it's going to be used server-side, the auth header should be the bot's token. If it's client-side, it should be the OAuth token. Simple

**Comment by beescuit** [comment]
Not a bad idea. It just needs to be something to authenticate the server for functionalities like balance modification (daily) or something like that

**Comment by beescuit** [comment]
Whoops.... Just remembered bladey-web doesn't use express or any "programmable" server-side part. I think the best way to do it then is to request the API directly from the browser and limit the endpoints so that they cant be abused (like instead of a "balance modification endpoint", we would have only a ".../daily" one). So the best idea is to use OAuth anyway. Also, answering @bolsomito even though this isn't a option: the key would be a environment var or something that wouldn't get influenced by the project being open source or not

**Comment by davipatury** [comment]
What about endpoints for personal text, background, etc., they need some kind of authentication, not only a limitation. @pedrofracassi suggestion works, but I don't think it's the cleanest one.

**Comment by beescuit** [comment]
Yeah, I mean it should be limited by user so that a token (via OAuth or whatever) points to a specific user that can only change their options (By limitations I meant account based ones and so on)

**Comment by dsgpaiva** [comment]
@beescuit you need to speak to me how the fucc would that work (like a PoC or something along those lines)

**Comment by beescuit** [comment]
Well, I don't really know. But maybe by using passport?

**Comment by davipatury** [comment]
@bolsomito According to @pedrofracassi, the Authentication header can be the user's OAuth token, the API uses the OAuth token to check if it's a valid token and to check which user is it. So, if you send a request to the personal text endpoint, it's going to change the OAuth token user's personal text.

But as I said it before, it's a little overkill because we'll need two OAuth connections (website and API), which could be prevented by not using a static website.

**Comment by dsgpaiva** [comment]
You can move `bladey-web` to other thingy, such as [express](https://expressjs.org) and you can host it at probably [Google](https://console.cloud.google.com) or [DigitalOcean](https://digitalocean.com), both of which I have free credit.

**Comment by dsgpaiva** [comment]
And while you're at it, you can also move `bladey-api`

**Comment by dsgpaiva** [comment]
and for CI, you can use [TeamCity](https://ci.c2g.space) or [Jenkins](https://jenkins.io)

**Comment by pedrofracassi** [comment]
Travis, hello?

**Comment by dsgpaiva** [comment]
Travis delivers Continuous Delivery for SSH enabled servers?

**Comment by pedrofracassi** [comment]
Nope

**Comment by dsgpaiva** [comment]
well, it [does](https://docs.travis-ci.com/user/deployment), but I still need to find for SSH. I've found for [Google App Engine](https://docs.travis-ci.com/user/deployment/google-app-engine/), but not for SSH or Google Compute Engine

**Comment by pedrofracassi** [comment]
![image](https://user-images.githubusercontent.com/25179120/45515133-aad9de00-b77d-11e8-8fe3-37a0e48214be.png)

**Comment by pedrofracassi** [comment]
I really don't think we need to move from the static website.

**Comment by dsgpaiva** [comment]
but it like, gives us more free room (at least from my pov) for doing "private" things 😈

**Comment by dsgpaiva** [comment]
> @bolsomito According to @pedrofracassi, the Authentication header can be the user's OAuth token, the API uses the OAuth token to check if it's a valid token and to check which user is it. So, if you send a request to the personal text endpoint, it's going to change the OAuth token user's personal text.
> 
> But as I said it before, it's a little overkill because we'll need two OAuth connections (website and API), which could be prevented by not using a static website.

**Comment by pedrofracassi** [comment]
What do you mean by "OAuth connections"? We'd use the user's token to make REST requests to Discord

**Comment by dsgpaiva** [comment]
```json
{
    "access_token": "6qrZcUqja7812RVdnEKjpzOL4CvHBFG",
    "token_type": "Bearer",
    "expires_in": 604800,
    "refresh_token": "D43f5y0ahjqew82jZ4NViEr2YafMKhue",
    "scope": "identify"
}
```
This is what Discord returns in /api/v6/oauth2/token. What do you use from it on both sides to validate that a real user is trying to do such thing?

[haha it's not real](https://discordapp.com/developers/docs/topics/oauth2#authorization-code-grant-access-token-response)

**Comment by pedrofracassi** [comment]
The `access_token`. We'd make a request using it to `/me` and make the changes to the user it returns.

**Comment by dsgpaiva** [comment]
umm that's the client secret token afaik

**Comment by dsgpaiva** [comment]
ok it's not, nvm

**Comment by dsgpaiva** [comment]
hmmm, that's a good idea, and in my eyes it's not *that* complicated to do that

**Comment by pedrofracassi** [comment]
We're already doing it for the Avatar endpoint and stuff (I think)

**Comment by dsgpaiva** [comment]
[discord.js](https://discord.js.org)

**Comment by dsgpaiva** [comment]
ah, you mean on web, nvm

**Comment by dsgpaiva** [comment]
ok, so this is what's needed to do:
+ establish a database connection on the API (through [mongoose](https://npmjs.com/package/mongoose) and [mongoose-findorcreate](https://npmjs.com/package/mongoose-findorcreate))
+ be able to change values from it (although that won't be complicated)
+ make an endpoint that will:
   + receive the user's access token & its changes
   + validate it by using `/users/@me` [wink wink](https://discordapp.com/developers/docs/resources/user#get-current-user)
   + if user is valid: 
      + grab its client ID
      + change the stuff that needs to be changed
      + save
   + else if user's invalid || no OAuth2 code || invalid changes:
      + send back a 403 error

**Comment by dsgpaiva** [comment]
at least on my head that works

**Comment by dsgpaiva** [comment]
i mean, tbh "mongoose-findorcreate" isn't required

**Comment by dsgpaiva** [comment]
or is it?

**Comment by dsgpaiva** [comment]
it is

**Comment by pedrofracassi** [comment]
Yeah
That's right

**Comment by pedrofracassi** [comment]
@davipatury

**Comment by davipatury** [comment]
When I said connections, I meant requests. It would require twice requests.

**Comment by dsgpaiva** [comment]
> meaned

![image](https://user-images.githubusercontent.com/25179120/45515133-aad9de00-b77d-11e8-8fe3-37a0e48214be.png)

**Comment by MrPowerGamerBR** [comment]
```
7 dias atrás "eu nun trabalho mais com o troca lâmina"
agora "kk eae men nintendo switch faca"
```
pill_irl

**Comment by dsgpaiva** [comment]
@MrPowerGamerBR 🤦‍♀️

**Comment by pedrofracassi** [comment]
@MrPowerGamerBR he still helps us

**Comment by dsgpaiva** [comment]
POST **ht​tp​s​:/​/​ap​i.​sw​itchb​l​ade.xy​z/d​at​​abase​/anything**
BODY: ```json
{ "property": "personalText", "value": "issa test lmaoooooooooooo", "oauth": "user-oauth-token" }
```
@pedrofracassi like this?

**Comment by pedrofracassi** [comment]
POST /users/0000000000000000
{
"personalText": "uiewrhuyvhgrjuvbhyujrtbhuiyhbj"
"sthElse": "rgsrfthrfyjhyjh"
}

**Comment by pedrofracassi** [comment]
And the token would be sent to the server through a header, not the request body

**Comment by dsgpaiva** [comment]
like "Authorization" ?

**Comment by davipatury** [comment]
It should be `/users/@me` instead of `/users/:id`, we can extract the ID from the token.

**Comment by pedrofracassi** [comment]
@davipatury nice, didn't think of that.

**Comment by Doges** [comment]
:thonk: ![](http://idonthave.life/i/514yQ.png)

**Comment by dsgpaiva** [comment]
switchblade_irl

**Comment by dsgpaiva** [comment]
*we discuss about it but we don't open a pull request for it*

**Comment by beescuit** [comment]
[umm.. why not?](https://github.com/nicholastay/passport-discord)

**Comment by dsgpaiva** [comment]
so, this is what me and @beescuit got made yesterday:

website sends a **GET** request to `https://api.switchblade.xyz/users/@me`:
  + user's OAuth2 stored in localStorage
  + parameter to change
  + value to be changed in the parameter

the API then does this:
  + requests Discord (`https://discordapp.com/api/v6/users/@me`) with authorization `Bearer OAUTH-TOKEN`
  + if Discord errors (for example a 401 unauthorized), send back a 401 Unauthorized / 403 Forbidden
  + else:
    + grab the client's id
    + change the parameter's value
    + save to database
    + return code 200 OK w/ `{ "success": true }`
    + **BUT** if in any of the above steps occurs an error:
      + return code 418 I'm a teapot / 406 Not Acceptable with `{ "success": false, "error": "error goes here" }`

**Comment by pedrofracassi** [comment]
Seems good.

**Comment by dsgpaiva** [comment]
This has already been closed in #10.
